### ISSUE TITLE
Running full suite fails with GC overhead limit exceeded

### ISSUE THREAD

**Issue by perNyfelt** [bedatadriven/renjin-release#5]
When running all tests it stop with a failure about "GC overhead limit exceeded"
Looking at reports on similar issues, e.g. https://github.com/gradle/gradle/issues/7746
it seems that this could probably be fixed by increasing the maxHeapSize for tests.

Here are the details:

vagrant@ubuntu-xenial:/home/ubuntu/renjin-release/packages$ ./gradlew test
Configuration on demand is an incubating feature.
Expiring Daemon because JVM heap space is exhausted
Daemon will be stopped at the end of the build after running out of JVM memory
Expiring Daemon because JVM heap space is exhausted
Expiring Daemon because JVM heap space is exhausted
Expiring Daemon because JVM heap space is exhausted
Expiring Daemon because JVM heap space is exhausted

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Could not determine the dependencies of task ':cran:fst:test'.
> Could not create task ':cran:fst:copyPackageResources'.
   > GC overhead limit exceeded

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* Where:
Script '/home/ubuntu/renjin-release/packages/gradle/report.gradle' line: 124

* What went wrong:
Failed to notify build listener.
> GC overhead limit exceeded

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.4.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 4m 21s

FAILURE: Build failed with an exception.

* What went wrong:
GC overhead limit exceeded

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.4.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 4m 35s
vagrant@ubuntu-xenial:/home/ubuntu/renjin-release/packages$

**Comment by akbertram** [comment]
Indeed, the GitHub action, which runs on a host with 7gb of memory, is also failing. It seems that Gradle doesn't get past the configuration stage, during which it builds a model of all the projects and tasks to execute. 

I know the full build runs on a machine with 32gb, but it will be most inconvenient if that were to be a minimum requirement. For one, it would rule out using GitHub's free hosted runners :-)

I suspect which we should start with upgrading to the latest version of Gradle, which promises some big improvements in multi-module builds. 

As a "quick" workaround, you can use the following script to avoid configuring the whole packages build:

```
#!/bin/sh -e

for pkg in `ls cran`
do
  ./gradlew cran:$pkg:test
done
```

**Comment by perNyfelt** [comment]
I have been running the build for a bit over one hour now with the following addition to packages/build.gradle:

subprojects {
  tasks.withType(Test) {
    maxHeapSize = "3g"
  }
}

**Comment by akbertram** [comment]
I believe that code above will only affect the TestTask, which runs Java tests. 

I suspect that having warmed up the caches will make a difference.

I'm taking a look now at the way the libstdc++ build is set up. It seems to pull in half of Renjin into the JVM running Gradle, which might have led to the OOM errors we're seeing during the "configuration" phase.

**Comment by perNyfelt** [comment]
OK interesting. I am still running by the way (1h 38min) so it does seem to help. Whether it will be enough or not remains to be seen. After a brief look, I think upgrading to the latest gradle is a good idea - it looks like it could improve things a lot.

**Comment by akbertram** [comment]
Great. I have a full build (minus tests) running on GitHub as well:
https://github.com/bedatadriven/renjin-release/runs/991101774

Fingers crossed!

**Comment by akbertram** [comment]
Adding the following to gradle.properties seems to resolve the problem:

```
org.gradle.jvmargs=-Xmx8G
```

At least when running on a VM with 30GB RAM.

This is a scan of the build:
https://scans.gradle.com/s/kpvebhdximr4a

Which might help us reduce memory consumption so that it can be run with a more modest memory profile.

**Comment by akbertram** [comment]
This seems to have been fixed by 042b5b9e86946c86380c70e7f3457a19364df498 and the full build is now running on th GitHub runner, which has a modest 8GB of memory.
